diskv.go

```diff
@@ -196,28 +204,18 @@ func (d *Diskv) WriteStream(key string, r io.Reader, sync bool) error {
 	return d.writeStreamWithLock(pathKey, r, sync)
 }
 
-// createKeyFileWithLock either creates the key file directly, or
-// creates a temporary file in TempDir if it is set.
+// createKeyFileWithLock creates the key file with a random extension. This
+// will be automatically renamed by writeStreamWithLock once the write has been
+// completed. This solves issue #63, where calling ReadStream, then updating the
+// key before reading completes, leads to the reader getting invalid data.
 func (d *Diskv) createKeyFileWithLock(pathKey *PathKey) (*os.File, error) {
-	if d.TempDir != "" {
-		if err := os.MkdirAll(d.TempDir, d.PathPerm); err != nil {
-			return nil, fmt.Errorf("temp mkdir: %s", err)
-		}
-		f, err := ioutil.TempFile(d.TempDir, "")
-		if err != nil {
-			return nil, fmt.Errorf("temp file: %s", err)
-		}
-
-		if err := os.Chmod(f.Name(), d.FilePerm); err != nil {
-			f.Close()           // error deliberately ignored
-			os.Remove(f.Name()) // error deliberately ignored
-			return nil, fmt.Errorf("chmod: %s", err)
-		}
-		return f, nil
-	}
-
-	mode := os.O_WRONLY | os.O_CREATE | os.O_TRUNC // overwrite if exists
-	f, err := os.OpenFile(d.completeFilename(pathKey), mode, d.FilePerm)
+	// Figure out the path and append a random number
+	path := fmt.Sprintf("%s.%d", d.completeFilename(pathKey), rand.Int())
+	// It's incredibly unlikely that the destination file will exist, but
+	// we want to be absolutely sure: O_EXCL means we'll get an error if the
+	// file already exists.
+	mode := os.O_WRONLY | os.O_CREATE | os.O_EXCL
+	f, err := os.OpenFile(path, mode, d.FilePerm)
 	if err != nil {
 		return nil, fmt.Errorf("open file: %s", err)
 	}
```

Comment:
Better to retry?

Reply:
I can implement a retry dance, but we are talking about collisions of 64-bit integers so it may be more reasonable to just print "go buy a lottery ticket" and shut things down for the day :)

Reply:
OK, how about we split the difference and use a bigger suffix :p

```go
buf := make([]byte, 16)
n, err := d.rand.Read(buf)
// handle err
path := fmt.Sprintf("%s.%x", d.completeFilename(pathKey), buf)
```

Reply:
Sounds agreeable to me